### PR TITLE
[Feature](bangpy-ops): Add new operator LogAddExp

### PR DESCRIPTION
**Pull request by SS7D631**
Thanks for your contribution and we appreciate it a lot. 

## 1. Motivation

logaddexp

## 2. Modification

Please briefly describe what modification is made in this pull request, and indicate where to make the modification.

## 3. Test Report

If you want to know how to do operator testing, you can see [GTest-User-Guide-zh](../docs/GTest-User-Guide-zh.md).

#### 3.1.1 Accuracy Acceptance Standard

For static threshold standard details, see: [MLU-OPS Accuracy Acceptance Standard](../docs/MLU-OPS精度验收标准.md).

- [ ] diff1, diff1 <= 3e-3
- [ ] diff2, diff2 <= 3e-3

#### 3.1.2 Operator Scheme checklist

|     No.      |           Details            |      Check Results       |
|----------------|---------------------------|---------------------|
|        1       |          Supported hardware         | MLU270 <br> MLU290 <br>MLU370|
|        2       |          Job types          |    block <br> U1 <br> U4    |
|        3       |         Layouts            |  NHWC 、NCHW、ARRAY etc    |
|        4       |         Whether multi-dimensions are supported              |                |
|        5       |          Whether element zero is supported             |               |
|        6       |         Data type(half/float)       |         half / float etc           |
|        7      |        Whether there is size limit           |             |

#### 3.1.3 New Feature Test

If you have checked the following items, please tick the relevant box.

- [ ] Data type test
- [ ] Multidimensional tensor test
- [ ] Layout test
- [ ] Different size/integer remainder end segment/alignment misalignment test
- [ ] Zero dimensional tensor test/zero element test
- [ ] stability test
- [ ] Multiple platform test
- [ ] Gen_case module test
- [ ] Nan/INF tests 
- [ ] For memory leak check details, see[GTest-User-Guide-zh](../docs/GTest-User-Guide-zh.md).
- [ ] For code coverage check details, see: [GTest-User-Guide-zh](../docs/GTest-User-Guide-zh.md).
- [ ] For I/O calculation efficiency check details, see: [MLU-OPS Performance Acceptance Criteria](../docs/MLU-OPS性能验收标准.md).

#### 3.1.4 Parameter Check

When a new operator is submitted, the test points are given and the test results are stated.

| Test Point         | Acceptance Criteria | Test Result (Error Message) |
| -------------- | -------- | -------------------- |
| Whether it conforms to the operator restriction | Normal error |                      |
| Whether illegal parameters are passed  | Normal error |                      |

### 3.2 Performance Test

See [MLU-OPS Performance Acceptance Criteria](../docs/MLU-OPS性能验收标准.md) for details.

Platform ：MLU270

|Operator|Mlu_hardware_time(us)|Mlu_interface_time(us)|Mlu_io_efficiency|Mlu_compute_efficiency|Mlu_workwpace_size(Bytes)|Data_type|Shape|
|-----|----|----|----|----|----|------|-----|
|op_name|   |    |     |    |    |    |     |
|op_name|   |    |     |    |    |    |     |

Platform ：MLU290

|Operator|Mlu_hardware_time(us)|Mlu_interface_time(us)|Mlu_io_efficiency|Mlu_compute_efficiency|Mlu_workwpace_size(Bytes)|Data_type|Shape|
|-----|----|----|----|----|----|------|-----|
|op_name|   |    |     |    |    |    |     |
|op_name|   |    |     |    |    |    |     |

Platform：MLU370

|Operator|Mlu_hardware_time(us)|Mlu_interface_time(us)|Mlu_io_efficiency|Mlu_compute_efficiency|Mlu_workwpace_size(Bytes)|Data_type|Shape|
|-----|----|----|----|----|----|------|-----|
|op_name|   |    |     |    |    |    |     |
|op_name|   |    |     |    |    |    |     |

### 3.3 Summary Analysis

Please give a brief overview here, if you want to note and summarize the content.

### CODE REVIEW

File: bangpy-ops/ops/logaddexp/logaddexp.py [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+# OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+# IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+# CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+# SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+# pylint: disable=missing-docstring, invalid-name, too-many-locals
+"""A multi-platform code link example test for BANGPy TCP."""
```
[the remark this answers is not in the document]
Comment (SS7D631) [reply bangpy-ops/ops/logaddexp/logaddexp.py:22]: 已修改

File: bangpy-ops/ops/logaddexp/logaddexp.py [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+            tcp.less_equal(bool_mark, input_buffer, threshold_value)
+
+    def mark_the_out_of_range_vlaue(
+        self,
+        input_buffer: ty.Buffer("nram"),
+        x: ty.Buffer("nram"),
+        y: ty.Buffer("nram")
+        ) -> None:
+        max_threshold = 10
+        min_threshold = -1 * 7.5
```
[the remark this answers is not in the document]
Comment (SS7D631) [reply bangpy-ops/ops/logaddexp/logaddexp.py:66]: 已修改

File: bangpy-ops/ops/logaddexp/test_logaddexp.py [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+# OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+# IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+# CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+# SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+# pylint: disable=missing-docstring, invalid-name, too-many-locals
+"""A multi-platform code link example test for BANGPy TCP."""
```
[the remark this answers is not in the document]
Comment (SS7D631) [reply bangpy-ops/ops/logaddexp/test_logaddexp.py:22]: 已修改

File: bangpy-ops/ops/logaddexp/test_logaddexp.py [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+        return
+
+    # origin input
+    sub_shape = shape[random.randint(0, len(shape) - 1):len(shape)]
+    data_x = np.random.uniform(low = -800, high = 800, size = shape)
+    data_y = np.random.uniform(low = -800, high = 800, size = sub_shape)
+    try:
+        mlu_ret = run(data_x, data_y, dtype)
+    except Exception as err:
+        print(str(err))
```
[the remark this answers is not in the document]
Comment (SS7D631) [reply bangpy-ops/ops/logaddexp/test_logaddexp.py:153]: 已修改

File: bangpy-ops/ops/logaddexp/logaddexp.py [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+DTYPES = [bangpy.float16, bangpy.float32]
+TARGET_LIST = ["mlu370-s4", "mlu220-m2", "mlu270", "mlu290"]
+KERNEL_NAME = "LogAddExp"
+
+
+class LogAddExp:
+    def __init__(self, cluster_num: ty.int32, dtype_bits: ty.int32, dtype: ty.string) -> None:
+        self.dtype = dtype
+        self.dtype_sz = dtype_bits
+        self.cluster_num = cluster_num
```
Comment (ShouyangDong) [inline bangpy-ops/ops/logaddexp/logaddexp.py:35]: 一样的道理，可以看看adjust_hue怎么写的

Comment (SS7D631) [reply bangpy-ops/ops/logaddexp/logaddexp.py]: 已修改

File: bangpy-ops/ops/logaddexp/logaddexp.py [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+        else:
+            tcp.less_equal(bool_mark, input_buffer, threshold_value)
+
+    def mark_the_out_of_range_vlaue(
+        self,
+        input_buffer: ty.Buffer("nram"),
+        x: ty.Buffer("nram"),
+        y: ty.Buffer("nram")
+        ) -> None:
+        max_threshold = 10
```
Comment (ShouyangDong) [inline bangpy-ops/ops/logaddexp/logaddexp.py:64]: 这个魔鬼数字注释说明一下吧

Comment (SS7D631) [reply bangpy-ops/ops/logaddexp/logaddexp.py]: 注释已加

**Review by ShouyangDong — APPROVED** [review]

**Pull request closed without merging**